k_thread_resume can cause k_sem_take with K_FOREVER to return -EAGAIN and crash

**Describe the bug**
When one thread is blocked on `k_sem_take` with `K_FOREVER` on a semaphore whose count is set to 0 and another thread calls `k_thread_resume` on it, the thread will resume and the `k_sem_take` call will return `EAGAIN`, an error code that is normallly reserved for the waiting period timing out. Zephyr then seems to crash.

**To Reproduce**
Steps to reproduce the behavior:
1. `cd samples/hello_world`
2. Replace `samples/hello_world/src/main.c` with the following:
```
#include <zephyr.h>
#include <sys/printk.h>

K_SEM_DEFINE(BLOCKSEMA, 0, 1);

void block_thread(void)
{
	printk("Pre sema take\n");
	
	int code = k_sem_take(&BLOCKSEMA, K_FOREVER);
	
	printk("Post sema take %i\n", code);
}

K_THREAD_DEFINE(block_thread_id, 512, (k_thread_entry_t)block_thread, NULL, NULL,
		NULL, 1, 0, 0);

void main(void)
{ 
	printk("Main\n");

	k_thread_resume(block_thread_id);
}
```
3. `west build -b qemu_x86 -t run .`
4. See error

**Expected behavior**
Either `k_sem_take` should continue blocking since `k_sem_give` wasn't called, or it should produce a different error code per [the documentation](https://docs.zephyrproject.org/latest/reference/kernel/synchronization/semaphores.html#c.k_sem_take). It should also not crash.

**Impact**
Unexpected behavior / crash confusing to users. 

**Logs and console output**
Output:
```
$ west build -b qemu_x86 -t run .
...
SeaBIOS (version rel-1.12.1-0-ga5cab58-dirty-20200625_115407-9426dddc0a1f-zephyr
)
Booting from ROM..*** Booting Zephyr OS build zephyr-v2.4.0-634-g53cc090c39a5  ***
Main
Pre sema take
Post sema take -11
FAILED: zephyr/CMakeFiles/run 
cd /home/herbert/zephyr-external/zephyr/samples/hello_world/build && /home/herbert/zephyr-sdk-0.11.4/sysroots/x86_64-pokysdk-linux/usr/bin/qemu-system-i386 -m 9 -cpu qemu32,+nx,+pae -device isa-debug-exit,iobase=0xf4,iosize=0x04 -no-reboot -nographic -net none -pidfile qemu.pid -chardev stdio,id=con,mux=on -serial chardev:con -mon chardev=con,mode=readline -icount shift=5,align=off,sleep=off -rtc clock=vm -kernel /home/herbert/zephyr-external/zephyr/samples/hello_world/build/zephyr/zephyr.elf
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /home/herbert/.local/bin/cmake --build /home/herbert/zephyr-external/zephyr/samples/hello_world/build --target run
```

**Environment:**
 - OS: Ubuntu 18.04
 - Board: qemu_x86
 - Toolchain: Zephyr SDK 0.11.4
 - Commit 53cc090

--- a/include/kernel.h
+++ b/include/kernel.h
@@ -644,6 +644,8 @@ __syscall int k_thread_join(struct k_thread *thread, k_timeout_t timeout);
  * This routine puts the current thread to sleep for @a duration,
  * specified as a k_timeout_t object.
  *
+ * @note if @a timeout is set to K_FOREVER then the thread is suspended.
+ *
  * @param timeout Desired duration of sleep.
  *
  * @return Zero if the requested time has elapsed or the number of milliseconds

--- a/kernel/include/ksched.h
+++ b/kernel/include/ksched.h
@@ -105,6 +105,11 @@ static inline bool z_is_idle_thread_object(struct k_thread *thread)
 #endif /* CONFIG_MULTITHREADING */
 }
 
+static inline bool z_is_thread_suspended(struct k_thread *thread)
+{
+	return (thread->base.thread_state & _THREAD_SUSPENDED) != 0U;
+}
+
 static inline bool z_is_thread_pending(struct k_thread *thread)
 {
 	return (thread->base.thread_state & _THREAD_PENDING) != 0U;

--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -1310,6 +1310,7 @@ int32_t z_impl_k_sleep(k_timeout_t timeout)
 	__ASSERT(!arch_is_in_isr(), "");
 	sys_trace_void(SYS_TRACE_ID_SLEEP);
 
+	/* in case of K_FOREVER, we suspend */
 	if (K_TIMEOUT_EQ(timeout, K_FOREVER)) {
 		k_thread_suspend(_current);
 		return (int32_t) K_TICKS_FOREVER;

--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -417,7 +417,10 @@ static void ready_thread(struct k_thread *thread)
 	__ASSERT_NO_MSG(arch_mem_coherent(thread));
 #endif
 
-	if (z_is_thread_ready(thread)) {
+	/* If thread is queued already, do not try and added it to the
+	 * run queue again
+	 */
+	if (!z_is_thread_queued(thread) && z_is_thread_ready(thread)) {
 		sys_trace_thread_ready(thread);
 		_priq_run_add(&_kernel.ready_q.runq, thread);
 		z_mark_thread_as_queued(thread);

--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -487,6 +487,12 @@ void z_impl_k_thread_resume(struct k_thread *thread)
 {
 	k_spinlock_key_t key = k_spin_lock(&sched_spinlock);
 
+	/* Do not try to resume a thread that was not suspended */
+	if (!z_is_thread_suspended(thread)) {
+		k_spin_unlock(&sched_spinlock, key);
+		return;
+	}
+
 	z_mark_thread_as_not_suspended(thread);
 	ready_thread(thread);
 

